lib/internal/modules/helpers.js

```diff
@@ -320,6 +329,28 @@ function normalizeReferrerURL(referrerName) {
 }
 
 
+let esmSyntaxErrorMessagesSet; // Declared lazily in shouldRetryAsESM
+let throwsOnlyInCommonJSErrorMessagesSet; // Declared lazily in shouldRetryAsESM
+/**
+ * After an attempt to parse a module as CommonJS throws an error, should we try again as ESM?
+ * @param {string} errorMessage The string message thrown by V8 when attempting to parse as CommonJS
+ * @param {string} source Module contents
+ */
+function shouldRetryAsESM(errorMessage, source) {
+  esmSyntaxErrorMessagesSet ??= new SafeSet(esmSyntaxErrorMessages);
+  if (esmSyntaxErrorMessagesSet.has(errorMessage)) {
+    return true;
+  }
+
+  throwsOnlyInCommonJSErrorMessagesSet ??= new SafeSet(throwsOnlyInCommonJSErrorMessages);
+  if (throwsOnlyInCommonJSErrorMessagesSet.has(errorMessage)) {
```

Comment:
nit: this could be simplified to:
```suggestion
  return throwsOnlyInCommonJSErrorMessagesSet.has(errorMessage) && contextifyShouldRetryAsESM(source);
```

Reply:
If it’s okay with you I think the current version makes the intent clearer, as the conditions of the two `if` statements are parallel with each other. I can’t imagine that there would be a performance difference between the two versions.